sklearn/linear_model/_logistic.py

```diff
@@ -233,7 +233,11 @@ def _logistic_grad_hess(w, X, y, alpha, sample_weight=None):
 
     def Hs(s):
         ret = np.empty_like(s)
-        ret[:n_features] = X.T.dot(dX.dot(s[:n_features]))
+        if sparse.issparse(X):
+            XTdX = safe_sparse_dot(X.T, dX)
+            ret[:n_features] = safe_sparse_dot(XTdX, s[:n_features])
```

Comment:
Would it make sense to try to replicate what `multi_dot` does internally? That is introspect the shape of `X`, `dX` and `s[:n_features]` to try to guess which order is the most cost-effective (both in terms of memory and CPU)?

Reply:
Yes, entirely! 🙂

I first wanted to fix failing tests by supporting sparse matrices; the next steps is to use the order with the smallest cost of operation. I am currently looking at the way matrices multiplications are implemented for sparse matrices within scipy.

With this use case here, maybe extending `numpy.linalg.multi_dot` to support sparse matrices would make sense?

Reply:
> With this use case here, maybe extending numpy.linalg.multi_dot to support sparse matrices would make sense?

numpy has no dependency / awareness of scipy. This could maybe contributed to `scipy.linalg.multi_dot`?

In any case we could have a `sklearn.utils.safe_multi_dot`.

Reply:
I see.

I will explore this; clearly porting `multi_dot` for sparse matrices is sound might be relevant for SciPy generally.